# Visualització de dades
# A7: Selecció del conjunt de dades
#### Per: Gisela Claret Tortajada

# 1. Introducció 

Aquest notebook ha estat creat per l'activitat 7 de l'assignatura Visualització de dades i s'hi carregaran i tractaran les dades seleccionades per crear un únic conjunt de dades per la realització de la Pràctica 2 (A9). 

Font de les dades: Les dades seleccionades són l'última versió del dataset "Lahman’s Baseball Database", que consta d'un total de 27 arxius .csv que contenen les estadístiques de la "Major League Baseball" (MLB) amb dades d'entre els anys 1871 fins el 2020. Així, les dades han estat extretes de: http://www.seanlahman.com/baseball-archive/statistics/ 

Per la realització de les visualitzacions demanades, s'ha escollit tenir solament en compte els pitchers, i, per tant, s'utilitzaran els següents arxius: "People.csv", "Pitching.csv", "Salaries.csv" i "Teams.csv"

Tant aquest notebook com la resta de la informació d'aquesta pràctica s'han inclòs en el GitHub següent: 

https://github.com/giselaclaret/PRACTICA_visualitzacio_de_dades

# 2. Càrrega de dades

In [1]:
import pandas as pd

In [2]:
#Carreguem els conjunts de dades:

df_people = pd.read_csv("./People.csv", keep_default_na=False, na_values=[''])
df_pitching = pd.read_csv("./Pitching.csv", keep_default_na=False, na_values=[''])
df_salaries = pd.read_csv("./Salaries.csv", keep_default_na=False, na_values=[''])
df_teams = pd.read_csv("./Teams.csv", keep_default_na=False, na_values=[''])

# 3. Neteja de dades

Netegem els conjunts de dades eliminant les columnes i dades que no ens interessen. 

### df_salaries

Comencem pel df Salaries. Aquest dataframe conté 26.428 entrades i 5 columnes diferents. Les dades són entre els anys 1985 i 2016, tal com veurem a continuació.

In [3]:
df_salaries.shape

(26428, 5)

In [4]:
df_salaries.head()

yearID teamID lgID   playerID  salary
0    1985    ATL   NL  barkele01  870000
1    1985    ATL   NL  bedrost01  550000
2    1985    ATL   NL  benedbr01  545000
3    1985    ATL   NL   campri01  633333
4    1985    ATL   NL  ceronri01  625000

In [5]:
df_salaries.sort_values(by=['yearID'], inplace=True)

In [6]:
print(df_salaries)

       yearID teamID lgID   playerID    salary
0        1985    ATL   NL  barkele01    870000
375      1985    OAK   AL  hendest01    150000
374      1985    OAK   AL  heathmi02    425000
373      1985    OAK   AL  griffal01    610000
372      1985    OAK   AL  davismi02    305000
...       ...    ...  ...        ...       ...
25862    2016    HOU   AL  altuvjo01   3687500
25863    2016    HOU   AL  castrja01   5000000
25864    2016    HOU   AL  correca01    516700
25866    2016    HOU   AL  feldmsc01   8000000
26427    2016    WAS   NL  zimmery01  14000000

[26428 rows x 5 columns]


In [7]:
df_salaries.columns

Index(['yearID', 'teamID', 'lgID', 'playerID', 'salary'], dtype='object')

D'aquest dataframe no fa falta mantenir la columna lgID, ja que es tracta de la lliga en la que juga l'equip i aquesta informació ja la obtindrem en el dataframe: df_teams.

In [8]:
del df_salaries['lgID']

In [9]:
df_salaries.head()

yearID teamID   playerID  salary
0      1985    ATL  barkele01  870000
375    1985    OAK  hendest01  150000
374    1985    OAK  heathmi02  425000
373    1985    OAK  griffal01  610000
372    1985    OAK  davismi02  305000

In [10]:
df_salaries.shape

(26428, 4)

### df_teams

En aquest dataframe es troba la informació dels equips de baseball. Tal com veurem, conté 2.955 entrades i consta de 48 columnes. Les dades es situen entre els anys 1871 i el 2017

In [11]:
df_teams.shape

(2955, 48)

In [12]:
df_teams.columns

Index(['yearID', 'lgID', 'teamID', 'franchID', 'divID', 'Rank', 'G', 'Ghome',
       'W', 'L', 'DivWin', 'WCWin', 'LgWin', 'WSWin', 'R', 'AB', 'H', '2B',
       '3B', 'HR', 'BB', 'SO', 'SB', 'CS', 'HBP', 'SF', 'RA', 'ER', 'ERA',
       'CG', 'SHO', 'SV', 'IPouts', 'HA', 'HRA', 'BBA', 'SOA', 'E', 'DP', 'FP',
       'name', 'park', 'attendance', 'BPF', 'PPF', 'teamIDBR',
       'teamIDlahman45', 'teamIDretro'],
      dtype='object')

In [13]:
df_teams.sort_values(by=['yearID'], inplace=True)

In [14]:
df_teams.head(1)

yearID lgID teamID franchID divID  Rank   G  Ghome   W   L  ...  DP     FP  \
0    1871   NA    BS1      BNA   NaN     3  31    NaN  20  10  ...  24  0.834   

                   name                 park  attendance  BPF  PPF  teamIDBR  \
0  Boston Red Stockings  South End Grounds I         NaN  103   98       BOS   

   teamIDlahman45  teamIDretro  
0             BS1          BS1  

[1 rows x 48 columns]

In [15]:
df_teams.tail(1)

yearID lgID teamID franchID divID  Rank   G  Ghome   W   L  ...  DP  \
2954    2020   NL    WAS      WSN     E     4  60   33.0  26  34  ...  48   

         FP                  name            park  attendance  BPF  PPF  \
2954  0.981  Washington Nationals  Nationals Park         0.0  103  102   

      teamIDBR  teamIDlahman45  teamIDretro  
2954       WSN             MON          WAS  

[1 rows x 48 columns]

In [16]:
df_teams = df_teams[['yearID','lgID', 'teamID', 'divID', 'Rank', 'W', 'L', 'name' ]]

In [17]:
df_teams.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2955 entries, 0 to 2954
Data columns (total 8 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   yearID  2955 non-null   int64 
 1   lgID    2955 non-null   object
 2   teamID  2955 non-null   object
 3   divID   1438 non-null   object
 4   Rank    2955 non-null   int64 
 5   W       2955 non-null   int64 
 6   L       2955 non-null   int64 
 7   name    2955 non-null   object
dtypes: int64(4), object(4)
memory usage: 207.8+ KB


Eliminem les dades anteriors a 1984 i posteriors a 2016

In [18]:
df_teams = df_teams[df_teams['yearID'] > 1984]  

In [19]:
df_teams = df_teams[df_teams['yearID'] < 2017]  

In [20]:
df_teams.head(3)

yearID lgID teamID divID  Rank   W   L                   name
1935    1985   NL    PHI     E     5  75  87  Philadelphia Phillies
1932    1985   AL    NYA     E     2  97  64       New York Yankees
1933    1985   NL    NYN     E     2  98  64          New York Mets

In [21]:
df_teams.tail(3)

yearID lgID teamID divID  Rank   W   L                name
2815    2016   AL    HOU     W     3  84  78      Houston Astros
2816    2016   AL    KCA     C     3  81  81  Kansas City Royals
2811    2016   NL    CIN     C     5  68  94     Cincinnati Reds

In [22]:
df_teams.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 918 entries, 1935 to 2811
Data columns (total 8 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   yearID  918 non-null    int64 
 1   lgID    918 non-null    object
 2   teamID  918 non-null    object
 3   divID   918 non-null    object
 4   Rank    918 non-null    int64 
 5   W       918 non-null    int64 
 6   L       918 non-null    int64 
 7   name    918 non-null    object
dtypes: int64(4), object(4)
memory usage: 64.5+ KB


Canviem els nombs de les columnes W, L i name per no confondre posteriorment amb les estadístiques dels jugadors.

In [23]:
df_teams.rename(columns = {'W' : 'team_W', 'L' : 'team_L', 'name' : 'team_name'}, inplace = True)

In [24]:
df_teams.columns

Index(['yearID', 'lgID', 'teamID', 'divID', 'Rank', 'team_W', 'team_L',
       'team_name'],
      dtype='object')

In [25]:
df_teams.shape

(918, 8)

Finalment al dataframe df_teams tenim 918 entrades i 8 columnes

### df_people

In [26]:
df_people.shape

(20153, 24)

In [27]:
df_people.columns

Index(['playerID', 'birthYear', 'birthMonth', 'birthDay', 'birthCountry',
       'birthState', 'birthCity', 'deathYear', 'deathMonth', 'deathDay',
       'deathCountry', 'deathState', 'deathCity', 'nameFirst', 'nameLast',
       'nameGiven', 'weight', 'height', 'bats', 'throws', 'debut', 'finalGame',
       'retroID', 'bbrefID'],
      dtype='object')

In [28]:
df_people.head(3)

playerID  birthYear  birthMonth  birthDay birthCountry birthState  \
0  aardsda01     1981.0        12.0      27.0          USA         CO   
1  aaronha01     1934.0         2.0       5.0          USA         AL   
2  aaronto01     1939.0         8.0       5.0          USA         AL   

  birthCity  deathYear  deathMonth  deathDay  ... nameLast    nameGiven  \
0    Denver        NaN         NaN       NaN  ...  Aardsma  David Allan   
1    Mobile     2021.0         1.0      22.0  ...    Aaron  Henry Louis   
2    Mobile     1984.0         8.0      16.0  ...    Aaron   Tommie Lee   

  weight height bats throws       debut   finalGame   retroID    bbrefID  
0  215.0   75.0    R      R  2004-04-06  2015-08-23  aardd001  aardsda01  
1  180.0   72.0    R      R  1954-04-13  1976-10-03  aaroh101  aaronha01  
2  190.0   75.0    R      R  1962-04-10  1971-09-26  aarot101  aaronto01  

[3 rows x 24 columns]

In [29]:
df_people = df_people[['playerID','birthCountry','nameFirst', 'nameLast', 'weight', 'height', 'bats', 'throws']]

In [30]:
df_people.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20153 entries, 0 to 20152
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   playerID      20153 non-null  object 
 1   birthCountry  20092 non-null  object 
 2   nameFirst     20116 non-null  object 
 3   nameLast      20153 non-null  object 
 4   weight        19337 non-null  float64
 5   height        19417 non-null  float64
 6   bats          18973 non-null  object 
 7   throws        19177 non-null  object 
dtypes: float64(2), object(6)
memory usage: 1.2+ MB


In [31]:
df_people.shape

(20153, 8)

Finalment al dataframe df_teams tenim 20153 entrades i 8 columnes

### df_pitching

Aquest dataframe presenta les estadístiques dels pitchers entre els anys 1871 i 2020. Conté un total de 48.399 entrades i 30 columnes. 

In [32]:
df_pitching.shape

(48399, 30)

In [33]:
df_pitching.columns

Index(['playerID', 'yearID', 'stint', 'teamID', 'lgID', 'W', 'L', 'G', 'GS',
       'CG', 'SHO', 'SV', 'IPouts', 'H', 'ER', 'HR', 'BB', 'SO', 'BAOpp',
       'ERA', 'IBB', 'WP', 'HBP', 'BK', 'BFP', 'GF', 'R', 'SH', 'SF', 'GIDP'],
      dtype='object')

In [34]:
df_pitching.sort_values(by=['yearID'], inplace=True)

In [35]:
df_pitching.head(1)

playerID  yearID  stint teamID lgID  W  L  G  GS  CG  ...  IBB  WP  HBP  \
0  bechtge01    1871      1    PH1   NA  1  2  3   3   2  ...  NaN   7  NaN   

   BK    BFP  GF   R  SH  SF  GIDP  
0   0  146.0   0  42 NaN NaN   NaN  

[1 rows x 30 columns]

In [36]:
df_pitching.tail(1)

playerID  yearID  stint teamID lgID  W  L   G  GS  CG  ...  IBB  WP  \
48398  zuberty01    2020      1    KCA   AL  1  2  23   0   0  ...  1.0   1   

       HBP  BK   BFP  GF   R   SH   SF  GIDP  
48398  1.0   0  99.0   8  11  1.0  1.0   3.0  

[1 rows x 30 columns]

Eliminem les files que no corresponen als anys entre 1984 i 2016

In [37]:
df_pitching = df_pitching[df_pitching['yearID'] > 1984]  

In [38]:
df_pitching = df_pitching[df_pitching['yearID'] < 2017]  

In [39]:
df_pitching.head(1)

playerID  yearID  stint teamID lgID  W  L   G  GS  CG  ...  IBB  WP  \
25282  powerte01    1985      1    CIN   NL  8  6  64   0   0  ...  8.0   1   

       HBP  BK    BFP  GF   R   SH   SF  GIDP  
25282  1.0   0  342.0  50  27  6.0  4.0   8.0  

[1 rows x 30 columns]

In [40]:
df_pitching.tail(1)

playerID  yearID  stint teamID lgID  W  L   G  GS  CG  ...  IBB  WP  \
44402  gileske01    2016      1    HOU   AL  2  5  69   0   0  ...  1.0  14   

       HBP  BK    BFP  GF   R   SH   SF  GIDP  
44402  2.0   0  286.0  24  32  2.0  1.0   0.0  

[1 rows x 30 columns]

In [41]:
df_pitching.shape

(19983, 30)

# 4. Combinació de df

Com que en els diferents dataframes tenim columnes en comú, combinarem els 5 dataframes per formar-ne un de sol.

Ens interessa la informació de solament els pitchers, per tant, començarem utilitzant el dataframe df_pitching.

In [42]:
df_pitching.columns

Index(['playerID', 'yearID', 'stint', 'teamID', 'lgID', 'W', 'L', 'G', 'GS',
       'CG', 'SHO', 'SV', 'IPouts', 'H', 'ER', 'HR', 'BB', 'SO', 'BAOpp',
       'ERA', 'IBB', 'WP', 'HBP', 'BK', 'BFP', 'GF', 'R', 'SH', 'SF', 'GIDP'],
      dtype='object')

In [43]:
df_people.columns

Index(['playerID', 'birthCountry', 'nameFirst', 'nameLast', 'weight', 'height',
       'bats', 'throws'],
      dtype='object')

In [44]:
df_final = df_pitching.merge(df_people, on = ['playerID'])

In [45]:
df_final.shape

(19983, 37)

In [46]:
df_final.columns

Index(['playerID', 'yearID', 'stint', 'teamID', 'lgID', 'W', 'L', 'G', 'GS',
       'CG', 'SHO', 'SV', 'IPouts', 'H', 'ER', 'HR', 'BB', 'SO', 'BAOpp',
       'ERA', 'IBB', 'WP', 'HBP', 'BK', 'BFP', 'GF', 'R', 'SH', 'SF', 'GIDP',
       'birthCountry', 'nameFirst', 'nameLast', 'weight', 'height', 'bats',
       'throws'],
      dtype='object')

In [47]:
df_final.head(5)

playerID  yearID  stint teamID lgID   W   L   G  GS  CG  ...   SH   SF  \
0  powerte01    1985      1    CIN   NL   8   6  64   0   0  ...  6.0  4.0   
1  powerte01    1986      1    CIN   NL  10   6  56  10   0  ...  9.0  6.0   
2  powerte01    1987      1    CIN   NL  10  13  34  34   2  ...  8.0  7.0   
3  powerte01    1988      2    DET   AL   1   1   4   2   0  ...  0.0  0.0   
4  powerte01    1988      1    KCA   AL   5   6  22  12   2  ...  2.0  4.0   

   GIDP  birthCountry  nameFirst  nameLast  weight  height  bats  throws  
0   8.0           USA        Ted     Power   215.0    76.0     R       R  
1  13.0           USA        Ted     Power   215.0    76.0     R       R  
2  12.0           USA        Ted     Power   215.0    76.0     R       R  
3   2.0           USA        Ted     Power   215.0    76.0     R       R  
4   6.0           USA        Ted     Power   215.0    76.0     R       R  

[5 rows x 37 columns]

In [48]:
df_salaries.shape

(26428, 4)

In [49]:
df_final = df_final.merge(df_salaries , how='left', on = ['playerID', 'yearID', 'teamID'])

In [50]:
df_final.shape

(19983, 38)

In [51]:
df_final = df_final[df_final['salary'].notna()]

In [52]:
df_final.shape

(11953, 38)

In [53]:
print(df_final.loc[df_final['playerID'] == 'degroja01'])

        playerID  yearID  stint teamID lgID   W  L   G  GS  CG  ...   SF  \
19356  degroja01    2015      1    NYN   NL  14  8  30  30   0  ...  7.0   
19357  degroja01    2016      1    NYN   NL   7  8  24  24   1  ...  3.0   

       GIDP  birthCountry  nameFirst  nameLast  weight  height  bats  throws  \
19356   8.0           USA      Jacob    deGrom   180.0    76.0     L       R   
19357  15.0           USA      Jacob    deGrom   180.0    76.0     L       R   

         salary  
19356  556875.0  
19357  607000.0  

[2 rows x 38 columns]


In [54]:
df_final.columns

Index(['playerID', 'yearID', 'stint', 'teamID', 'lgID', 'W', 'L', 'G', 'GS',
       'CG', 'SHO', 'SV', 'IPouts', 'H', 'ER', 'HR', 'BB', 'SO', 'BAOpp',
       'ERA', 'IBB', 'WP', 'HBP', 'BK', 'BFP', 'GF', 'R', 'SH', 'SF', 'GIDP',
       'birthCountry', 'nameFirst', 'nameLast', 'weight', 'height', 'bats',
       'throws', 'salary'],
      dtype='object')

In [55]:
df_teams.columns

Index(['yearID', 'lgID', 'teamID', 'divID', 'Rank', 'team_W', 'team_L',
       'team_name'],
      dtype='object')

In [56]:
df_final = df_final.merge(df_teams , on = [ 'yearID', 'lgID', 'teamID'])

In [57]:
df_final.columns

Index(['playerID', 'yearID', 'stint', 'teamID', 'lgID', 'W', 'L', 'G', 'GS',
       'CG', 'SHO', 'SV', 'IPouts', 'H', 'ER', 'HR', 'BB', 'SO', 'BAOpp',
       'ERA', 'IBB', 'WP', 'HBP', 'BK', 'BFP', 'GF', 'R', 'SH', 'SF', 'GIDP',
       'birthCountry', 'nameFirst', 'nameLast', 'weight', 'height', 'bats',
       'throws', 'salary', 'divID', 'Rank', 'team_W', 'team_L', 'team_name'],
      dtype='object')

In [58]:
df_final.shape

(11953, 43)

Afegim la columna continent de naixement

In [59]:
df_final['birthCountry'].value_counts()

USA               9534
D.R.               948
Venezuela          333
Mexico             228
P.R.               202
CAN                152
Japan              132
Cuba                97
Panama              64
Australia           45
South Korea         40
Nicaragua           32
Germany             30
Taiwan              20
United Kingdom      20
Aruba               14
Colombia            13
Curacao             12
Viet Nam            10
Netherlands          9
Jamaica              8
Afghanistan          3
Philippines          3
France               1
Saudi Arabia         1
Indonesia            1
Spain                1
Name: birthCountry, dtype: int64

In [60]:
def label_continent (row):
    if row['birthCountry'] == 'USA':
        return 'North America'
    if row['birthCountry'] == 'D.R.':
        return 'Central America'
    if row['birthCountry'] == 'Venezuela':
        return 'South America'
    if row['birthCountry'] == 'Mexico':
        return 'Central America'
    if row['birthCountry'] == 'P.R.':
        return 'Central America'
    if row['birthCountry'] == 'CAN':
        return 'North America'
    if row['birthCountry'] == 'Japan':
        return 'Asia'
    if row['birthCountry'] == 'Cuba':
        return 'Central America'
    if row['birthCountry'] == 'Panama':
        return 'Central America'
    if row['birthCountry'] == 'Australia':
        return 'Oceania'
    if row['birthCountry'] == 'South Korea':
        return 'Asia'
    if row['birthCountry'] == 'Nicaragua':
        return 'Central America'
    if row['birthCountry'] == 'Germany':
        return 'Europe'
    if row['birthCountry'] == 'Taiwan':
        return 'Asia'
    if row['birthCountry'] == 'United Kingdom':
        return 'Europe'
    if row['birthCountry'] == 'Aruba':
        return 'South America'
    if row['birthCountry'] == 'Colombia':
        return 'South America'
    if row['birthCountry'] == 'Curacao':
        return 'South America'
    if row['birthCountry'] == 'Viet Nam':
        return 'Asia'
    if row['birthCountry'] == 'Netherlands':
        return 'Europe'
    if row['birthCountry'] == 'Jamaica':
        return 'Central America'
    if row['birthCountry'] == 'Afghanistan':
        return 'Asia'
    if row['birthCountry'] == 'Philippines':
        return 'Asia'
    if row['birthCountry'] == 'France':
        return 'Europe'
    if row['birthCountry'] == 'Saudi Arabia':
        return 'Asia'
    if row['birthCountry'] == 'Spain':
        return 'Europe'
    if row['birthCountry'] == 'Indonesia':
        return 'Asia'
    else:
        return 'Other'

In [61]:
df_final.apply (lambda row: label_continent(row), axis=1)
df_final['birth_continent'] = df_final.apply (lambda row: label_continent(row), axis=1)

In [62]:
df_final['birth_continent'].value_counts()

North America      9686
Central America    1579
South America       372
Asia                210
Europe               61
Oceania              45
Name: birth_continent, dtype: int64

Afegim una columna amb l'index de massa corporal del jugador

In [63]:
df_final.weight

0        215.0
1        220.0
2        188.0
3        185.0
4        185.0
         ...  
11948    225.0
11949    225.0
11950    205.0
11951    215.0
11952    190.0
Name: weight, Length: 11953, dtype: float64

In [64]:
df_final['IMC_jugador'] = (df_final.weight * 703) / (df_final.height ** 2)

In [65]:
df_final.columns

Index(['playerID', 'yearID', 'stint', 'teamID', 'lgID', 'W', 'L', 'G', 'GS',
       'CG', 'SHO', 'SV', 'IPouts', 'H', 'ER', 'HR', 'BB', 'SO', 'BAOpp',
       'ERA', 'IBB', 'WP', 'HBP', 'BK', 'BFP', 'GF', 'R', 'SH', 'SF', 'GIDP',
       'birthCountry', 'nameFirst', 'nameLast', 'weight', 'height', 'bats',
       'throws', 'salary', 'divID', 'Rank', 'team_W', 'team_L', 'team_name',
       'birth_continent', 'IMC_jugador'],
      dtype='object')

In [66]:
df_final.shape

(11953, 45)

In [67]:
df_final.to_excel(r'baseball_dataset.xlsx', index = False)